README.md

# Morse Code Over Packet Protocol (MOPP) - UDP

## Introduction

MOPP is a protocol designed for the transmission of Morse code over data protocols that use payloads consisting of 8-bit bytes. It is suitable for both LoRa and IP networks, where byte strings are transmitted as packet payloads.

This protocol description outlines the structure of MOPP packets, the encoding of Morse code elements, and how the protocol can be used over UDP for IP networks.

The current document describes version 1 of the protocol.

## Design Goals

- Allow faithful transmission of Morse code.
- Support multiple languages and characters.
- Include information about the original speed.
- Be efficient to accommodate bandwidth constraints.
- Avoid the use of Zero bytes to prevent misinterpretation.
- Use UDP as the transport layer for IP networks.

## Protocol Structure

### Packet Structure:

- 2 bits: Protocol Version (01b for version 1)
- 6 bits: Serial Number
- 6 bits: Speed at origin in WpM (Words per Minute, 5 - 60 decimal)

### Payload:

The payload consists of Morse code information encoded as follows:

- 2 bits for each Morse code element:
  - 01b: dit
  - 10b: dah
  - 00b: End of Character (EOC)
  - 11b: End of Word (EOW)


#### Heartbeat (Modification for UDP Chat Server)

The MOPP UDP Chat server is always trying to receive packets. If nothing is sent, the connection will timeout.
This timeout triggers the sending of empty hearbeat packets to all clients (no payload).
If the clients respond, their session gets updated.


### Complete Packet Format:

```
+----------------------+-------------------+
| Protocol Version     | Serial Number     |
+----------------------+-------------------+
| Speed at Origin (WpM)| Payload           |
+----------------------+-------------------+
```

## Example:

Let's consider the Morse code for the word "PARIS" at a speed of 16 WpM:

- Protocol Version: 01
- Serial Number: 011011
- Speed: 010000
- 'P' (di dah dah dit EOC): 01 10 10 01 00
- 'A' (di dah EOC): 01 10 00
- 'R' (di dah dit EOC): 01 10 01 00
- 'I' (di dit EOC): 01 01 00
- 'S' (di di dit EOW): 01 01 01 11

Packing into a byte string:

`01011011 01000001 10100100 01100001 10010001 01000101 01110000`

In hexadecimal notation:

`5B 41 A4 61 91 45 70`

As an ASCII string:

```
[A¤a`Ep
```

## UDP Transmission:

- Use UDP for packet transmission.
- Send the MOPP packet as the payload of the UDP packet.
- Define a port for MOPP communication (standard 7373).

## Conclusion

MOPP over UDP provides an efficient and language-agnostic method for transmitting Morse code over IP networks, 
catering to the constraints of bandwidth and ensuring faithful representation of Morse code messages.

## Implementations
- Python (PyPI): [PyPI Package MOPP](https://test.pypi.org/project/mopp/)

## References
- [Original MOPP Protocol description](https://github.com/oe1wkl/Morserino-32/blob/master/Documentation/Protocol%20Description/morse_code_over_packet_protocol.md) ([Local copy](./doc/protocol-mopp.orig.md)) by Willi, OE1WKL
- Encoder and Decoder in Python: implementation taken from [Sigbit](https://github.com/tuxintrouble/sigbit) by Sebastian Stetter, DJ5SE. Other parts have been derived from the implementation by SP9WPN.


CHANGELOG.md
1.0.9 - set initial speed timings
1.0.8 - add duration json

wip/test.py

my_array = [60, -60, 180, -60, 380, -60]
my_v = [-2000, +50, -50, +50, -50, +50, -50, +150]
my_test_bug = [ -347, +145, -183, +51, -160, +51, -47, +47, -51, +47, -211, +144 ] 

def splitter(my_array):
    print (my_array)

    first_two_elements = my_array[:2]
    positive_elements = [abs(element) for element in first_two_elements]
    smallest_value = min(positive_elements)

    print (smallest_value)

    fractional_multiples = []
    non_fractional_multiples = []

    i = 0
    for element in my_array:
        if element % smallest_value == 0:
            fractional_multiples.append(element)
            i += 1
        else:
            break

    non_fractional_multiples = my_array[i:]
    print (fractional_multiples)
    print (non_fractional_multiples)



#splitter(my_array=my_array)

#splitter(my_array=my_v)
#splitter(my_array=my_test_bug)

def normalize_duration_timings(duration_array=my_test_bug):
    # Duration array representing Morse code tones (with imperfect timing)

    # Find the position of the first non-pause element
    first_non_pause_index = next((i for i, duration in enumerate(duration_array) if duration > 0), None)

    positive_durations = [duration for duration in duration_array if duration > 0]
    average_duration = float(sum(positive_durations))/float(len(positive_durations)) 

    # Estimate words per minute and calculate dit duration
    average_dit_duration = sum([duration for duration in duration_array[first_non_pause_index:] if duration > 0]) / len([duration for duration in duration_array[first_non_pause_index:] if duration > 0])
    wpm_estimate = int(1200 / average_dit_duration) if average_dit_duration > 0 else 0
    dit_length_estimate = int(1200 / wpm_estimate)
    dah_length_estimate = int(3*dit_length_estimate)

    # Define a threshold to distinguish between Dits and Dahs
    threshold = average_duration # 3 * average_dit_duration 
    #print (threshold)

    # Convert the array to Morse code string
    morse_code = ""
    normalized_durations = []
    for duration in duration_array[first_non_pause_index:]:
        if duration > 0 and duration <= threshold:  # Dit
            morse_code += "."
            normalized_durations.append(dit_length_estimate)
            normalized_durations.append(-dit_length_estimate)
        elif duration >= threshold:  # Dah
            morse_code += "-"
            normalized_durations.append(dah_length_estimate)
            normalized_durations.append(-dit_length_estimate)
        elif duration < 0 and abs(duration) >= threshold: # Pause
            morse_code += " " # FIXME: more types of pauses (word, character, ...)

    result = {"wpm_estimate": wpm_estimate, "morse_code_normalized": morse_code, "normalized_durations": normalized_durations}

    return result


#print (normalize_duration_timings(duration_array=my_v))
#print (normalize_duration_timings(duration_array=my_test_bug))



def wpm_estimater (duration_array):
    my_test_bug = [ -347, +145, -183, +51, -160, +51, -47, +47, -51, +47, -211, +144 ] 
    duration_array = my_test_bug
    positive_durations = [duration for duration in duration_array if duration > 0]
    average_duration = float(sum(positive_durations))/float(len(positive_durations)) 
    threshold = average_duration
    possible_dits = [duration for duration in positive_durations if duration < threshold]
    possible_dahs = [duration for duration in positive_durations if duration >= threshold]
    print (positive_durations)
    print (possible_dits)
    print (possible_dahs)
    average_dit_duration = ( sum (possible_dits ) + 1/3.0 * sum(possible_dahs) ) / len (positive_durations)
    wpm_estimate = int(1200 / average_dit_duration) if average_dit_duration > 0 else 0
    print (average_dit_duration)
    print (wpm_estimate)

python/testing/test_chat_server.py
#!/usr/bin/python3
import socket
import mopp

class ChatClient:
    wpm = 23
    bufferSize = 1024
    UDPClientSocket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)

    def __init__(self, wpm=23, ip="127.0.0.1", port=7373):
        self.m = mopp.mopp(wpm=wpm)
        self.serverAddressPort = (ip, port)

    def send_receive(self, msg):
        f=self.m.str2mopp(msg)
        bytesToSend = str.encode(f)
        self.UDPClientSocket.sendto(bytesToSend, self.serverAddressPort)
        msgFromServer = self.UDPClientSocket.recvfrom(self.bufferSize)
        mm = msgFromServer[0].decode("utf-8", errors='ignore')

        m1 = self.m.decodePacket(mm)
        m2 = self.m.mopp2str(mm)
        print (self.m.rx_serial, self.m.rx_wpm, m1, m2)

#c = ChatClient()
c = ChatClient(ip="49.12.102.144")
c.send_receive("hi")
c.send_receive("hello")
c.send_receive("bye")


python/mopp/tests/test_chat_server.py
#!/usr/bin/python3
import socket
import python.mopp.src.mopp_protocol.mopp as mopp

class ChatClient:
    wpm = 23
    bufferSize = 1024
    UDPClientSocket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)

    def __init__(self, wpm=23, ip="127.0.0.1", port=7373):
        self.m = mopp.mopp(wpm=wpm)
        self.serverAddressPort = (ip, port)

    def send_receive(self, msg):
        f=self.m.str2mopp(msg)
        bytesToSend = str.encode(f)
        self.UDPClientSocket.sendto(bytesToSend, self.serverAddressPort)
        msgFromServer = self.UDPClientSocket.recvfrom(self.bufferSize)
        mm = msgFromServer[0].decode("utf-8", errors='ignore')

        m1 = self.m.decodePacket(mm)
        m2 = self.m.mopp2str(mm)
        print (self.m.rx_serial, self.m.rx_wpm, m1, m2)

#c = ChatClient()
c = ChatClient(ip="49.12.102.144")
c.send_receive("hi")
c.send_receive("hello")
c.send_receive("bye")


javascript/mopp.js
function pad(num, size) { // Ref: https://stackoverflow.com/questions/2998784/how-to-output-numbers-with-leading-zeros-in-javascript
  var s = "000000" + num;
  return s.substr(s.length-size);
}

function dec2bin(dec) { // Ref: https://stackoverflow.com/questions/9939760/how-do-i-convert-an-integer-to-binary-in-javascript
  a = (dec >>> 0).toString(2);
  b = pad(a, 6);
  return b;
}

String.prototype.leftJustify = function( length, char ) { // Ref: https://gist.github.com/biesiad/889139
  var fill = [];
  while ( fill.length + this.length < length ) {
    fill[fill.length] = char;
  }
  return fill.join('') + this;
}


class mopp {
  constructor(wpm = 20) {
    this.tx_wpm = wpm;
    this.tx_serial = 1;
    this.tx_protocol_version = 1;
    this.morse = {
      "0": "-----",
      "1": ".----",
      "2": "..---",
      "3": "...--",
      "4": "....-",
      "5": ".....",
      "6": "-....",
      "7": "--...",
      "8": "---..",
      "9": "----.",
      "a": ".-",
      "b": "-...",
      "c": "-.-.",
      "d": "-..",
      "e": ".",
      "f": "..-.",
      "g": "--.",
      "h": "....",
      "i": "..",
      "j": ".---",
      "k": "-.-",
      "l": ".-..",
      "m": "--",
      "n": "-.",
      "o": "---",
      "p": ".--.",
      "q": "--.-",
      "r": ".-.",
      "s": "...",
      "t": "-",
      "u": "..-",
      "v": "...-",
      "w": ".--",
      "x": "-..-",
      "y": "-.--",
      "z": "--..",
      "=": "-...-",
      "/": "-..-.",
      "+": ".-.-.",
      "-": "-....-",
      ".": ".-.-.-",
      ",": "--..--",
      "?": "..--..",
      ":": "---...",
      "!": "-.-.--",
      "'": ".----.",
      ";": "-.-.-.",
      "&": ".-...",
      "@": ".--.-.",
      "\u00e4": ".-.-",
      "\u00f6": "---.",
      "\u00fc": "..--",
      "ch": "----",
      "\"": ".-..-.",
      "(": "-.--.",
      ")": "-.--.-",
      "<sk>": "...-.-",
      "<bk>": "-...-.-",
      " ": ""
    };
  }

  str2mopp(str) {
    var m;
    m = this.tx_protocol_version;
    m += dec2bin(this.tx_serial);
    m += dec2bin(this.tx_wpm);

    for (var c, _pj_c = 0, _pj_a = str, _pj_b = _pj_a.length; _pj_c < _pj_b; _pj_c += 1) {
      c = _pj_a[_pj_c];

      if (c === " ") {
        m += "11";
        continue;
      }

      for (var b, _pj_f = 0, _pj_d = this.morse[c.toLowerCase()], _pj_e = _pj_d.length; _pj_f < _pj_e; _pj_f += 1) {
        b = _pj_d[_pj_f];

        if (b === ".") {
          m += "01";
        } else {
          m += "10";
        }
      }

      m += "00";
    }

    m = m.slice(0, -2) + "11";
    m = m.leftJustify(Number.parseInt(8 * Math.ceil(m.length / 8.0)), "0");
    this.tx_serial += 1;
    return this.bit2str(m);
  }

  bit2str(bit) {
    var m, res;
    m = bit;
    res = "";

    for (var i = 0, _pj_a = m.length; i < _pj_a; i += 8) {
      res += String.fromCharCode(Number.parseInt(m.slice(i, i + 8), 2));
    }

    return res;
  }

  decodePacket(packet) {
    /*
    //// byte 1: header; first two bits are the protocol version (curently 01), plus 6 bit packet serial number (starting from random)
    //// byte 2: first 6 bits are wpm (must be between 5 and 60; values 00 - 04 and 61 to 63 are invalid), the remaining 2 bits are already data payload!
     */
    var bits, l, msg;
    bits = 0;
    l = bits.length;
    this.rx_protocol_version = bits.slice(0, 2);
    this.rx_serial = bits.slice(3, 8);
    this.rx_wpm = Number.parseInt(bits.slice(9, 14), 2);
    msg = "";

    for (var i = 14, _pj_a = l; i < _pj_a; i += 2) {
      if (bits.slice(i, i + 2) === "01") {
        msg += ".";
      } else {
        if (bits.slice(i, i + 2) === "10") {
          msg += "-";
        } else {
          if (bits.slice(i, i + 2) === "00") {
            msg += " ";
          } else {
            if (bits.slice(i, i + 2) === "11") {
              msg += " ";
            }
          }
        }
      }
    }

    return msg;
  }

  mopp2str(packet) {
    var klist, msg, p, v, vlist;
    p = this.decodePacket(packet);
    klist = list(this.morse.keys());
    vlist = list(this.morse.values());
    msg = "";

    for (var c, _pj_c = 0, _pj_a = p.split(" "), _pj_b = _pj_a.length; _pj_c < _pj_b; _pj_c += 1) {
      c = _pj_a[_pj_c];
      v = list(this.morse.keys())[list(this.morse.values()).index(c)];
      msg += v;
    }

    return msg;
  }

}

/*
_pj.set_properties(mopp, {
  "morse": 
  "rx_protocol_version": "01",
  "rx_serial": 1,
  "tx_protocol_version": "01",
  "tx_serial": 1,
  "tx_wpm": 20,
  "tx_wpm": 20
});

*/

python/mopp/src/mopp_protocol/mopp.py
#!/usr/bin/python3
from math import ceil

class mopp():
    tx_serial = 1
    tx_protocol_version = '01'
    tx_wpm = 20

    rx_serial = 1
    rx_protocol_version = '01'
    tx_wpm = 20

    morse = {
        "0" : "-----", "1" : ".----", "2" : "..---", "3" : "...--", "4" : "....-", "5" : ".....",
        "6" : "-....", "7" : "--...", "8" : "---..", "9" : "----.",
        "a" : ".-", "b" : "-...", "c" : "-.-.", "d" : "-..", "e" : ".", "f" : "..-.", "g" : "--.",
        "h" : "....", "i" : "..", "j" : ".---", "k" : "-.-", "l" : ".-..", "m" : "--", "n" : "-.",
        "o" : "---", "p" : ".--.", "q" : "--.-", "r" : ".-.", "s" : "...", "t" : "-", "u" : "..-",
        "v" : "...-", "w" : ".--", "x" : "-..-", "y" : "-.--", "z" : "--..", "=" : "-...-",
        "/" : "-..-.", "+" : ".-.-.", "-" : "-....-", "." : ".-.-.-", "," : "--..--", "?" : "..--..",
        ":" : "---...", "!" : "-.-.--", "'" : ".----.", ";" : "-.-.-.", "&" : ".-...", "@" : ".--.-.",
        "ä" : ".-.-", "ö" : "---.", "ü" : "..--", "ch" : "----", '"' : ".-..-.", "(" : "-.--.", ")" : "-.--.-",
        "<sk>" : "...-.-", "<bk>" : "-...-.-",
        " " : "" # for eow spacing
    }

    def __init__(self, wpm=20):
        self.tx_wpm = wpm

    def str2mopp(self, str):
        # Header
        m = self.tx_protocol_version               
        m += bin(self.tx_serial)[2:].zfill(6)
        m += bin(self.tx_wpm)[2:].zfill(6)

        # Payload
        for c in str:
            if c == " ":
                m += '11'			# spaces not supported by morserino - treat as word
                continue

            for b in self.morse[c.lower()]:
                if b == '.':
                    m += '01'       # .
                else:
                    m += '10'       # -

            m += '00'				# EOC

        m = m[0:-2] + '11'	        # final EOW
 
        m = m.ljust(int(8*ceil(len(m)/8.0)),'0') # fill rest with 0
        self.tx_serial += 1
        print (m)

        return self.bit2str(m)

    def bit2str (self, bit):
        # Encode data to binary notation
        m = bit
        res = ''
        for i in range (0, len(m), 8):
            #debug: print (m[i:i+8])
            res += chr(int(m[i:i+8],2))
        
        return res

    def decodePacket(self, packet):
        """
        //// byte 1: header; first two bits are the protocol version (curently 01), plus 6 bit packet serial number (starting from random)
        //// byte 2: first 6 bits are wpm (must be between 5 and 60; values 00 - 04 and 61 to 63 are invalid), the remaining 2 bits are already data payload!

        """
        bits = "".join("{:08b}".format(ord(c)) for c in packet)
        # debug print (bits)

        l = len(bits)

        # Extract Header
        self.rx_protocol_version = bits[0:2]
        self.rx_serial = bits[3:8]
        self.rx_wpm = int(bits[9:14],2)

        # Extract payload
        msg = ""
        for i in range (14, l, 2):
            if bits[i:i+2] == "01": # .
                msg += "."
            elif bits[i:i+2] == "10": # -
                msg += "-"
            elif bits[i:i+2] == "00": # EOC
                msg += " "
            elif bits[i:i+2] == "11": # EOW
                msg += " "
            
        #debug print (self.rx_protocol_version, self.rx_serial, self.rx_wpm, msg)
        return (msg)

    def str2hex(self,str):
        return ":".join("{:02x}".format(ord(c)) for c in str)
    
    def str2bit(self,str):
        return ":".join("{:08b}".format(ord(c)) for c in str)

    def mopp2str (self, packet):
        p = self.decodePacket(packet)
        
        klist = list(self.morse.keys())
        vlist = list(self.morse.values())

        msg = ""
        for c in p.split(" "):
            #thing_index = thing_list.index(elem) if elem in thing_list else -1
            #v = "*"
            #if c in vlist:
            #    v = klist[vlist.index(c)]
            v = list(self.morse.keys())[list(self.morse.values()).index(c)]
            msg += v

        # debug print (msg)
        return (msg)


#m = mopp(wpm=23)
#f=m.str2mopp("Gerolf ok")
#print(m.str2bit(f))
#m.decodePacket(f)
#m.mopp2str(f)


# FIXME: utf8 encoding
# TODO: merge with sigbit code

python/mopp/src/mopp/__init__.py
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-
# This file is part of the SigBit project
# https://github.com/tuxintrouble/sigbit
# Author: Sebastian Stetter, DJ5SE
# License: GNU GENERAL PUBLIC LICENSE Version 3
#
# common utility functions for SigBit
from math import ceil
import json

def ditlen(wpm):
    """returns the lenght of a dit in seconds for a given words-per-minute"""
    #PARIS incl. Abstände == 50 ditlängen -> 1 ditlänge at1wpm = 60s / (50 * wpm)
    return 60 / (50 * wpm)


###morse LUT###
morse = {
	"0" : "-----", "1" : ".----", "2" : "..---", "3" : "...--", "4" : "....-", "5" : ".....",
	"6" : "-....", "7" : "--...", "8" : "---..", "9" : "----.",
	"a" : ".-", "b" : "-...", "c" : "-.-.", "d" : "-..", "e" : ".", "f" : "..-.", "g" : "--.",
	"h" : "....", "i" : "..", "j" : ".---", "k" : "-.-", "l" : ".-..", "m" : "--", "n" : "-.",
	"o" : "---", "p" : ".--.", "q" : "--.-", "r" : ".-.", "s" : "...", "t" : "-", "u" : "..-",
	"v" : "...-", "w" : ".--", "x" : "-..-", "y" : "-.--", "z" : "--..", "=" : "-...-",
	"/" : "-..-.", "+" : ".-.-.", "-" : "-....-", "." : ".-.-.-", "," : "--..--", "?" : "..--..",
	":" : "---...", "!" : "-.-.--", "'" : ".----.", ";" : "-.-.-.", "&" : ".-...", "@" : ".--.-.",
        "ä" : ".-.-", "ö" : "---.", "ü" : "..--", "ch" : "----", '"' : ".-..-.", "(" : "-.--.", ")" : "-.--.-",
        "<sk>" : "...-.-", "<bk>" : "-...-.-"
  }


def encode(text): # FIXME: Make a first function to convert character to .- language and a 2nd to convert to 01010 and a final one to convert to mopp binary
    """takes a string of characters and returns a wordbuffer"""
    wordbuffer = []
    for c in text:
        if c == " ":
            wordbuffer.append("11")
        if c in morse:
            for el in morse[c]:
                if el == "-":
                    wordbuffer.append("10")
                if el == ".":
                    wordbuffer.append("01")
            wordbuffer.append("00")
    if len(wordbuffer) > 0:
        wordbuffer.pop()
    wordbuffer.append("11")
    return wordbuffer
                

def decode(buffer):
    """takes a wordbuffer and returns a string of characters"""
    global morse    
    outcode = ""
    outchars = ""
    for el in buffer:
        if el == "01":
            outcode += "."
        elif el == "10":
            outcode += "-"
        elif el == "00":
            
            for letter, code in morse.items():
                if code == outcode:
                    outchars += letter
                    outcode = ""
        elif el =="11":
            for letter, code in morse.items():
                if code == outcode:
                    outchars += letter
                    outcode = ""
                    outchars = outchars + " "
                    return outchars


#make own zfill for uPython
def zfill(str,digits):
    '''we need to implement our own zfill for uPython)'''
    if len(str)>=digits:
        return str
    else:
        return ((digits - len(str)) * '0')+str


#make own ljust for uPython
def ljust(string, width, fillchar=' '):
    '''returns the str left justified, remaining space to the right is filed with fillchar, if str is shorter then width, original string is returned '''
    while len(string) < width:
        string += fillchar
    return string



# Module for MOPP protocol
# Taken and adjusted from the m32-chat-server implementation by SP9WPN

class Mopp:
    serial = 1

    def __init__(self, speed = 20):
        self.speed = speed
        self.set_speed (speed=speed)
        return
    
    def set_speed(self, speed = 20):
        # Ref timings: https://morsecode.world/international/timing.html#:~:text=It's%20clear%20that%20this%20makes,%22)%20which%20also%20makes%20sense.
        speed_wpm = speed
        self.dit_duration = int(60 / (50*speed_wpm)*1000)
        self.dah_duration = 3*self.dit_duration
        self.eoc_duration = 3*self.dit_duration
        self.eow_duration = 7*self.dit_duration

    def _str2hex(self, bytes):
        hex = ":".join("{:02x}".format(c) for c in bytes)
        return hex

    def _str2bin(self, bytes):
        bincode = "".join("{:08b}".format(c) for c in bytes)
        return bincode

    def mopp(self, speed, msg):
        #logging.debug("Encoding message with "+str(speed)+" wpm :"+str(msg))

        morse = {
            "0" : "-----", "1" : ".----", "2" : "..---", "3" : "...--", "4" : "....-", "5" : ".....",
            "6" : "-....", "7" : "--...", "8" : "---..", "9" : "----.",
            "a" : ".-", "b" : "-...", "c" : "-.-.", "d" : "-..", "e" : ".", "f" : "..-.", "g" : "--.",
            "h" : "....", "i" : "..", "j" : ".---", "k" : "-.-", "l" : ".-..", "m" : "--", "n" : "-.",
            "o" : "---", "p" : ".--.", "q" : "--.-", "r" : ".-.", "s" : "...", "t" : "-", "u" : "..-",
            "v" : "...-", "w" : ".--", "x" : "-..-", "y" : "-.--", "z" : "--..", "=" : "-...-",
            "/" : "-..-.", "+" : ".-.-.", "-" : "-....-", "." : ".-.-.-", "," : "--..--", "?" : "..--..",
            ":" : "---...", "!" : "-.-.--", "'" : ".----."
        }

        m = '01'				# protocol version
        m += bin(self.serial)[2:].zfill(6)
        m += bin(speed)[2:].zfill(6)

        for c in msg:
            if c == " ":
                continue				# spaces not supported by morserino!

            for b in morse[c.lower()]:
                if b == '.':
                    m += '01'
                else:
                    m += '10'

            m += '00'				# EOC

        m = m[0:-2] + '11'			# final EOW
        m = m.ljust(int(8*ceil(len(m)/8.0)),'0')

        #print (m, " ENCODER") # FIXME

        res = ''
        for i in range (0, len(m), 8):
            #print (m[i:i+8], bytes(chr(int(m[i:i+8],2)),"latin_1"), i, " ENCODER")
            res += chr(int(m[i:i+8],2))

        self.serial += 1
        return bytes(res,'latin_1') # WATCH OUT: UNICODE MAKES MULTI-BYTES 

    def _stripheader(self, msg):
        res = bytes(0x00) + bytes(msg[1] & 3) + msg[2:]
        return res

    def msg_strcmp (self, data_bytes, speed, msg):
        if self._stripheader(data_bytes) == self._stripheader(self.mopp(speed, msg)):
            return True
        else:
            return False
        
    def received_speed (self, data_bytes): # FIXME
        speed = data_bytes[1] >> 2 
        return speed

    def received_serial (self, data_bytes): # FIXME
        #myserial = data_bytes[0] >> 2 
        myserial = 0 # FIXME
        return myserial

    def received_data (self, data_bytes): # TODO
        return self._str2hex(data_bytes)

    def decode_message (self, data_bytes):
        if len(data_bytes) < 1:
            return {"Keepalive": True}
        
        speed = data_bytes[1] >> 2 

        # Convert symbols to string of 0 and 1 again
        n = ""
        for l in [data_bytes[i:i+1] for i in range(len(data_bytes))]:
            n += "{:08b}".format(ord(l))
        
        # list of bit pairs 01, 10, 11, 00
        sym = [n[i:i+2] for i in range(0, len(n), 2)] 
        protocol = sym[0]
        serial = int("".join(sym[1:4]),2)

        # Extract message in format ./-/EOC/EOW
        msg = ""
        for i in range (14, len(n), 2):
            s = n[i:i+2]
            msg += self._mopp2morse(s)

        return {"Protocol": protocol, "Serial": serial, "Speed": speed, "Message": msg}


    def _mopp2morse(self, sym):
        s = ""
        if sym == '01':
            s = '.'
        elif sym == '10':
            s = '-'
        elif sym == '00':
            s = 'EOC'
        elif sym == '11':
            s = 'EOW'
        else:
            return ""
            #logging.debug ("This should not happen: symbol ", s)
        return s
    
    def _morse2txt(self, morse):
        return
    
    def return_duration_json(self, message):
        json_string = '{"durations": []}'
        data = json.loads(json_string)

        for symbol in message:
            if symbol == ".":
                data['durations'].append(self.dit_duration)
                data['durations'].append(-self.dit_duration)
            elif symbol == "-":
                data['durations'].append(self.dah_duration)
                data['durations'].append(-self.dit_duration)
            elif symbol == "C": # EOC
                data['durations'].append(-self.eoc_duration)
            elif symbol == "W": # EOW
                data['durations'].append(-self.eow_duration)

        updated_json_string = json.dumps(data, indent=2)
        return updated_json_string